talisker/logs.py

```diff
@@ -326,46 +327,67 @@ def format(self, record):
     def clean_message(self, s):
         return s.replace('"', '\\"').replace('\n', '\\n')
 
-    def remove_quotes(self, s):
-        return s.replace('"', '')
-
     def logfmt(self, structured):
-        logfmt = (self.logfmt_atom(*kv) for kv in structured.items())
-        return " ".join(l for l in logfmt if l is not None)
+        formatted = (
+            '{}={}'.format(k, v) for k, v in self.logfmt_atoms(structured))
+        return " ".join(formatted)
+
+    def logfmt_atoms(self, structured):
+        for k, v in structured.items():
+            key = None
+            try:
+                if v is not False and not v:
+                    continue
+
+                key = self.logfmt_key(k)
+
+                if isinstance(v, dict):
+                    for k2, v2 in v.items():
+                        yield (
+                            key + '_' + self.logfmt_key(k2),
+                            self.logfmt_value(v2)
+                        )
+                else:
+                    yield key, self.logfmt_value(v)
+            except Exception as e:
+                # an exception has occured during logging
+                # TODO: send to sentry?
+                if key is not None:
+                    yield key, '"???"'
+
+    def logfmt_key(self, k):
+        if isinstance(k, bytes):
+            k = k.decode('utf8')
 
-    def logfmt_atom(self, k, v):
-        if not v:
-            return None
+        # TODO: look at optimizing this to avoid string allocations
 
-        size = self.MAX_VALUE_SIZE
+        k = k[:self.MAX_KEY_SIZE].strip()
```

Comment:
We should `.strip()` before truncating, so that leading whitespace won't cause the key to be unnecessarily truncated